doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+# Reference Counting
+
+## Problem
+
+Elektra's data structures are shared between multiple owners.
+A `Key` can be owned by many different `KeySet`s.
+Both `Key` and `KeySet` can be used in bindings.
+If the languages of those bindings have automatic memory management, this also implies a shared ownership.
+
+To solve this issue of shared ownership, we need reference counting.
+
+An adjacent problem is that the names of `Key`s must be read-only while the `Key` is part of a `KeySet`.
+This can also be solved by a secondary counter just for `KeySet`s.
+
+There has also been [discussion](copy_on_write.md) about adding further data structures, which would need reference counters as well.
+
+## Constraints
+
+- Deleting data structures that still have references must be prevented
+- The APIs for `ksAppendKey` and `ksNew` must remain ergonomic.
+  The reference counting must not prevent the inline creation of a `Key` while adding it to a `KeySet`.
+
+## Assumptions
+
+- The reference counter doesn't need to be thread-safe.
+- A reference counter of size `uint16_t` is large enough for all use cases.
+  Further restricting the counter by a small amount to allow for error value is also acceptable.
```

Comment:
The current C++ binding increments the reference counter for usage of `Key` in blocks. So in the case of recursion, `uint16_t` might not be enough.

I agree that we should assume some fixed size to be enough, maybe even `uint16_t` (and fix the C++ Binding).

Reply:
I mean we could go all the way and just use `size_t`, then it's physically impossible to have more than that.

Reply:
> The current C++ binding increments the reference counter for usage of Key in blocks. So in the case of recursion, `uint16_t` might not be enough.

Do you mean something like this:

```cpp
void foo(Key k, int depth) {
   if (depth > 0) {
      foo(k);
   }
}
```

Because that can be solved without even touching the binding (I think):

```cpp
void foo(Key & k, int depth) {
   if (depth > 0) {
      foo(k);
   }
}
```

In any case, IMO assuming `uint16_t` for now is fine. We can always change it to e.g. `uint32_t` without much trouble.

Reply:
> I mean we could go all the way and just use `size_t`, then it's physically impossible to have more than that.

I think `uint32_t` should be more than enough. I wouldn't go to `size_t`, because that is also the maximum alignment. So having a `size_t refs` and e.g. `int flags` in the same struct has more padding than with `uint32_t refs` or `uint16_t refs`.

Reply:
> Do you mean something like this:

Yes.

> Because that can be solved without even touching the binding (I think):

People might use call by value `Key` and still expect this to work. Call by value is now much more popular due to move semantics.

> I think uint32_t should be more than enough

Having fixed-size types has the advantages:

- same behavior on all systems
- mmap files are more likely to be compatible

But then we need to document how big arrays etc. can be. Maybe even its own little decision.

Reply:
> due to move semantics

A move constructor shouldn't change the refcount, because it just _moves_ the reference. If it's not certain that a move constructor would be used, then (especially for a deeply recursive function) pass-by-value should not be used for stack size reasons alone.

However, I've already said that we can discuss a larger refcount in future. But the size of the field has basically no impact on the decision. We should not discuss it now. All I wanted to say with the assumption (and maybe I'll rephrase it) is that (1) `uint16_t` is the absolute minimum and (2) for most or even all cases, it also good enough.

Reply:
> However, I've already said that we can discuss a larger refcount in future.

I don't think there is any sensible way to change anything in refcounting after 1.0.

Reply:
Well, if there is no `keyGetRefCount()` function then I don't see an issue. Changing the type of the refcount would be a purely internal change with no affect on the API. It would still change behaviour, so the question is how strict we want to be, but I'd argue it would just be a minor `x.1.0` change and not a major breaking change.

If there is a `keyGetRefCount()` then changing the return type would be a breaking change. But we could technically do a hacky solution like:

```c
uint16_t keyGetRefCount(Key * key) {
    if (key->refcount > UINT16_MAX) return UINT16_MAX;
    return key->refcount;
}

uint32_t keyGetRefCountLarge(Key * key) {
    return key->refcount;
}
```

All this is to say: For the IMO _highly unlikely_ case that `uint16_t` turns out to be too small, there are ways to go forward. So IMHO using `uint16_t` is fine. If it is possible without increasing the struct sizes, we could also use `uint32_t`. But I really wouldn't sacrifices extra bytes for every `struct _Key`, just of the hypothetical case where `uint16_t` isn't enough.